### PR TITLE
feat(query-insights): demote score for wasteful single-field bitmap indexes

### PR DESCRIPTION
**Pull request by tnaum-ms**
## Summary

Adds a score demotion rule to the Query Insights static analysis for single-field bitmap indexes that return a large fraction of the collection.

Previously, a query using a bitmap index with high selectivity (≥20% of documents returned) could still receive an `Excellent` performance rating because the index was technically efficient — `docsExamined ≈ nReturned`. This was misleading: bitmap indexes on low-cardinality fields carry ongoing write and storage overhead that outweighs the marginal read benefit when the index returns many documents.

### What changed

- **Demotion rule**: When `isBitmap: true`, the index is single-field, and selectivity ≥ 20%, the bitmap badge type is set to `negative` and the performance score is demoted one level (e.g. `Excellent` → `Good`).
- **Compound indexes are excluded**: The penalty only applies to single-field indexes. A compound index that _includes_ a bitmap field on a low-cardinality key still provides meaningful selectivity via its other keys.
- **Spec updated**: `query-insights-static-analysis.md` documents the demotion table and design decision.
- **Tests**: 4 new test cases added to `ExplainPlanAnalyzer.addIndexStrategyAdvisories.test.ts`; all 10 bitmap tests pass.

### Non-DocumentDB safety

`isBitmap` is a DocumentDB-specific explain plan field. Standard MongoDB explain plans never include it, so this rule is never triggered for non-DocumentDB clusters.

### CODE REVIEW

File: src/documentdb/queryInsights/ExplainPlanAnalyzer.ts [hunk 1]
```diff
@@ -912,21 +912,59 @@ export class ExplainPlanAnalyzer {
...
+                const executionStages = (explainResult.executionStats as Document | undefined)?.executionStages as
+                    | Document
+                    | undefined;
+                const ixscanExec = this.findStageInPlan(executionStages, 'IXSCAN');
+                const indexUsage = ixscanExec?.indexUsage as Array<{ scanKeys?: string[] }> | undefined;
+                const isSingleField = indexUsage?.some((u) => u.scanKeys && u.scanKeys.length === 1) ?? false;
+
+                const coverage =
+                    totalCollectionDocs && totalCollectionDocs > 0
+                        ? analysis.nReturned / totalCollectionDocs
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply src/documentdb/queryInsights/ExplainPlanAnalyzer.ts]: Fixed in 365bb252 — added `Math.min(..., 1)` clamp to match the earlier coverage-badge logic. This prevents the details string from showing >100% when `totalCollectionDocs` is an underestimate of `nReturned`.

File: src/documentdb/queryInsights/query-insights-static-analysis.md [hunk 2]
```diff
@@ -227,9 +227,25 @@ Fires when the `queryPlanner.winningPlan` IXSCAN stage has `isBitmap === true`.
...
 efficiency, unlike the heuristic-based low-cardinality badge.
 
-| diagnosticId   | Type    | Message      | Score effect |
-| -------------- | ------- | ------------ | ------------ |
-| `bitmap_index` | neutral | Bitmap index | None         |
+| diagnosticId   | Condition                                          | Type     | Message      | Score effect      |
+| -------------- | -------------------------------------------------- | -------- | ------------ | ----------------- |
+| `bitmap_index` | Bitmap, compound index OR selectivity < 20%        | neutral  | Bitmap index | None              |
+| `bitmap_index` | Bitmap, single-field index AND selectivity >= 20%  | negative | Bitmap index | Demoted one level |
+
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply src/documentdb/queryInsights/query-insights-static-analysis.md]: Fixed in 3e847d1b — updated both the spec markdown (line 211) and the source JSDoc on `addIndexStrategyAdvisories()` to clarify that bitmap index and severe multikey expansion advisories can demote the score. The wording now reads: "Most are informational, but selected advisories can also demote the score when they identify a clear index-strategy problem."

**Comment by tnaum-ms** [comment]
**Review fix #3: `isSingleField` can misclassify mixed `indexUsage` entries**

Fixed in 12127d37 — changed `isSingleField` from `.some()` to a strict check requiring exactly one `indexUsage` entry with exactly one `scanKey`. If DocumentDB returns multiple `indexUsage` entries with mixed key counts, the index is conservatively treated as compound to avoid false demotions.

Added a test (`does not demote score when indexUsage has mixed single/multi-key entries`) to lock down this edge case.

**Comment by tnaum-ms** [comment]
**Review fix #4: Planner IXSCAN and execution-stats IXSCAN are not correlated**

Fixed in bb5a4274 — the bitmap check now correlates the planner IXSCAN and the execution-stats IXSCAN by `indexName`. If the first IXSCAN found in each tree refers to a different index (e.g., OR plans, index intersection), the single-field detection falls back to neutral instead of potentially inspecting a different index's `scanKeys`.

Added a test (`does not demote when planner and exec IXSCAN have different index names`) to cover this edge case.

File: src/documentdb/queryInsights/ExplainPlanAnalyzer.ts [hunk 3]
```diff
@@ -912,21 +912,59 @@ export class ExplainPlanAnalyzer {
...
+                    const scoreOrder: Array<ExecutionStatsAnalysis['performanceRating']['score']> = [
+                        'excellent',
+                        'good',
+                        'fair',
+                        'poor',
+                    ];
+                    const currentIndex = scoreOrder.indexOf(analysis.performanceRating.score);
+                    if (currentIndex >= 0 && currentIndex < scoreOrder.length - 1) {
+                        analysis.performanceRating.score = scoreOrder[currentIndex + 1];
+                    }
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply src/documentdb/queryInsights/ExplainPlanAnalyzer.ts]: Fixed in 3b49ace1 — extracted a `demoteScoreOneLevel(analysis)` private static helper and a module-level `SCORE_ORDER` constant. Both the bitmap and multikey demotion blocks now use the shared helper, eliminating the duplicated pattern.

File: src/documentdb/queryInsights/ExplainPlanAnalyzer.addIndexStrategyAdvisories.test.ts [hunk 4]
```diff
@@ -53,10 +64,7 @@ function makeExplainResult(options: { isBitmap?: boolean; indexName?: string } =
             totalKeysExamined: 100,
             executionStages: {
                 stage: 'FETCH',
-                inputStage: {
-                    stage: 'IXSCAN',
-                    indexName: options.indexName ?? 'someIndex_1',
-                },
+                inputStage: execIxscan,
             },
         },
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply src/documentdb/queryInsights/ExplainPlanAnalyzer.addIndexStrategyAdvisories.test.ts:77]: Fixed in 5ec272a8 — `makeExplainResult` now accepts an `nReturned` option so `executionStats.nReturned` stays aligned with the `analysis.nReturned` override used in each test. All test call sites that vary `analysis.nReturned` now pass the same value to the fixture.

**Comment by tnaum-ms** [comment]
**Review fix #7: Double-demotion behavior is not specified**

Fixed in a229c1eb — added a test (`demotes score twice when both bitmap and severe multikey thresholds are met`) confirming that advisory demotions are cumulative: if both `bitmap_index` and `severe_multikey_expansion` fire, the score is reduced by two levels (e.g., Excellent → Fair).

Also documented this in the static-analysis spec under a new "Cumulative demotions" note.

File: src/documentdb/queryInsights/ExplainPlanAnalyzer.ts [hunk 5]
```diff
@@ -912,21 +912,59 @@ export class ExplainPlanAnalyzer {
...
-                    ),
+                    details: shouldDemote
+                        ? l10n.t(
+                              'The database used a bitmap index on a low-cardinality field.\n\nThis single-field index splits the collection into very few buckets, returning {0}% of documents. The ongoing write and storage cost on every insert and update outweighs the marginal read benefit.\n\nConsider hiding this index if no other queries depend on it.',
+                              ((coverage ?? 0) * 100).toFixed(1),
+                          )
+                        : l10n.t(
+                              'The database used a bitmap index to execute this query.\n\nBitmap indexes are an internal optimization that DocumentDB applies to low-cardinality fields (fields with few distinct values, such as booleans or status codes). They are space-efficient but less selective than B-tree indexes on high-cardinality fields.\n\nThis is expected behavior and does not indicate a problem. If query performance is a concern, consider filtering on a more selective field or using a compound index.',
+                          ),
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply src/documentdb/queryInsights/ExplainPlanAnalyzer.ts:977]: Already resolved — `l10n/bundle.l10n.json` was updated in commit ebc3b3cd (the `chore: l10n, prettier` commit) and already contains the new bitmap demotion details string. No additional action needed.

**Review by guanzhousongmicrosoft — APPROVED** [review]
